### PR TITLE
Add kubernetes config & hacks to so that we can autoscale!

### PR DESCRIPTION
**Pull request by kshahat**
1. Created k8s manifest for deploy/svc
2. changed the makefile to use mvn in online mode and to skip the tests [which fixes a known issue.](https://github.com/spotify/bigtable-autoscaler/issues/104)
3. changed the dockerfile to use `java` instead of `/usr/bin/java` [which fixes an issue.](https://github.com/spotify/bigtable-autoscaler/issues/64)

TODO:
CircleCi support

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
-NS = spotify
+NS = default
 
 SERVICE_NAME = bigtable-autoscaler
-SERVICE_IMAGE_VERSION = latest
+SERVICE_IMAGE_VERSION ?= 0.0.1
```
Comment (dangermike) [inline Makefile:4]: What does `?=` mean here?

Comment (kshahat) [reply Makefile:4]: use `0.0.1` if the value isn't already defined

File: bigtable-autoscaler-service.yaml [hunk 2]
```diff
@@ -0,0 +1,81 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: bigtable-autoscaler
+  annotations:
+    cloud.google.com/neg: '{"ingress": true}'
+  labels:
+    app: bigtable-autoscaler
+spec:
```
Comment (dangermike) [inline bigtable-autoscaler-service.yaml:7]: This does not require an ingress -- we will never expose this service outside the cluster

File: bigtable-autoscaler-service.yaml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  template:
+    metadata:
+      labels:
+        app: bigtable-autoscaler 
+      annotations:
+        cluster-autoscaler.kubernetes.io/safe-to-evict: 'true'
+    spec:
+      containers:
+      - name: bigtable-autoscaler 
+        image: us.gcr.io/triggeredmail/bigtable-autoscaler:0.0.1
```
Comment (dangermike) [inline bigtable-autoscaler-service.yaml]: To use the deploy tool, I'd rather have `{{image}}` here.

File: Makefile [hunk 4]
```diff
@@ -1,7 +1,7 @@
-NS = spotify
+NS = default
 
 SERVICE_NAME = bigtable-autoscaler
-SERVICE_IMAGE_VERSION = latest
```
Comment (alexagriffith) [inline Makefile:4]: jw, why not use latest?

File: bigtable-autoscaler-service.yaml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      containers:
+      - name: bigtable-autoscaler 
+        image: us.gcr.io/triggeredmail/bigtable-autoscaler:0.0.1 
+        imagePullPolicy: Always
+        volumeMounts:
+          - mountPath: /service-account
+            name: service-account
+        resources:
+          requests:
+            cpu: 1
```
Comment (alexagriffith) [inline bigtable-autoscaler-service.yaml]: also curious, why are we requesting a whole cpu?

**Review by alexagriffith — APPROVED** [review]

**Review by dangermike — CHANGES_REQUESTED** [review]
I'd like to have this building with CircleCI and deploying using the deploy tool. That means:

* have the image in the Dockerfile be a templated value as `{{image}}`
* have the `Makefile` put on an additional docker tag with the git SHA
   * see code below from [here](https://github.com/TriggerMail/recommendationsvc2/blob/master/cloudbuild#L6-L20)). 
* create .circleci/config.yml that can build and push the dockerfile

You don't have to use cloudbuilder to have an image end up in GCR, but I don't want images built on our laptops to be used in production. Airflow doing that is enough.


```bash
if [[ ${CIRCLE_SHA1+x} ]]; then
    if [ "${CIRCLE_BRANCH}" != "master" ]; then
        echo "Branch ${CIRCLE_BRANCH} != master. Exiting."
        exit 0
    fi
    echo "${GCR_TM_AUTH}" > "${HOME}/gcloud-service-key.json"
    gcloud auth activate-service-account --key-file="${HOME}/gcloud-service-key.json"

    GIT_SHA=$CIRCLE_SHA1
    GCLOUD_KEY_PARAM=" --gcloud-key-path ${HOME}/gcloud-service-key.json"

else
    GIT_SHA=$(git rev-parse HEAD)
    GCLOUD_KEY_PARAM=""
fi
```

File: Makefile (hunk 1 above)
Comment (dangermike) [reply Makefile:4]: TIL. Thanks!

**Review by dangermike — APPROVED** [review]
Let's put the circle and deploy tool stuff in another PR. This looks pretty good for now, though Alexa's right: you don't need a whole CPU :)

**Comment by alexagriffith** [comment]
> I'd like to have this building with CircleCI and deploying using the deploy tool. That means:

for this to work you also need to go to settings in the repo and change the branch rules. create a rule named `master` and one of the settings is circleCI config

File: bigtable-autoscaler-service.yaml (hunk 2 above)
Comment (kshahat) [reply bigtable-autoscaler-service.yaml:7]: good point - didn't consider the ingress

File: bigtable-autoscaler-service.yaml (hunk 3 above)
Comment (kshahat) [reply bigtable-autoscaler-service.yaml]: yeah forgot to add

File: Makefile (hunk 4 above)
Comment (kshahat) [reply Makefile:4]: wouldn't know which image you're using as the running image will always be 'latest'

File: bigtable-autoscaler-service.yaml (hunk 5 above)
Comment (kshahat) [reply bigtable-autoscaler-service.yaml]: forgot to change it 🤷‍♂️

File: Makefile (hunk 4 above)
Comment (alexagriffith) [reply Makefile:4]: Yeah but the most updated image would always be the latest one right? And you want to use the most recent one? Idk maybe hard coding the version is better in this case

Comment (kshahat) [reply Makefile:4]: yes but without a deploy the image on a pod will also be latest. It could be two different images but both tagged as `latest`

Comment (kshahat) [reply Makefile:4]: version numbers allow you to refer to a specific PR that reflects the changes that are running

**Comment by shraykay** [comment]
driveby comment: when this becomes a real thing you may want to tune the jmx flags and also enable jmx metrics and funnel them into stackdriver

**Pull request merged**